Sprint 16

Lagowazzz opened the pull request.

Lagowazzz committed:
OnboardingVC was created with PageVC and works

Lagowazzz committed:
The Onboarding screen is shown only on first launch

Lagowazzz committed:
CategoryViewModel was created
CategoryVC -> MVVM

Helvetii951 commented on Trackers/CoreData/Store/AppSettingsStore.swift:
```diff
@@ -0,0 +1,37 @@
+        }
+    }
+    
+    func setHasSeenOnboarding(_ seen: Bool) {
+        let request: NSFetchRequest<AppSettings> = AppSettings.fetchRequest()
+        request.fetchLimit = 1
+        
+        do {
+            let settings = try context.fetch(request).first ?? AppSettings(context: context)
+            settings.hasSeenOnboarding = seen
+            try context.save()
+        } catch {
+            assertionFailure("Failed to update AppSettings: \(error)")
+        }
+    }
```
Обычно для таких легковесных настроек/флагов использую `UserDefaults`. Такой подход будет полегче по ресурсам.

Lagowazzz replied:
Специально сделал через CoreData, чтобы было больше практики)

Helvetii951 commented on Trackers/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,149 @@
+        super.init(coder: coder)
+    }
+    
+    private lazy var pages: [UIViewController] = {
+        let red = UIViewController()
+        let redImage = UIImage(named: "red")
+        let redImageView = UIImageView(image: redImage)
+        redImageView.contentMode = .scaleAspectFill
+        red.view.addSubview(redImageView)
+        redImageView.translatesAutoresizingMaskIntoConstraints = false
+        redImageView.leadingAnchor.constraint(equalTo: red.view.leadingAnchor).isActive = true
+        redImageView.trailingAnchor.constraint(equalTo: red.view.trailingAnchor).isActive = true
+        redImageView.bottomAnchor.constraint(equalTo: red.view.bottomAnchor).isActive = true
+        redImageView.topAnchor.constraint(equalTo: red.view.topAnchor).isActive = true
+        
+        let redLabel = UILabel()
+        redLabel.text = "Даже если это\n не литры воды и йога"
+        redLabel.font = .boldSystemFont(ofSize: 32)
+        redLabel.textColor = .black
+        redLabel.textAlignment = .center
+        redLabel.numberOfLines = 3
+        red.view.addSubview(redLabel)
+        redLabel.translatesAutoresizingMaskIntoConstraints = false
+        redLabel.leadingAnchor.constraint(equalTo: red.view.leadingAnchor, constant: 16).isActive = true
+        redLabel.trailingAnchor.constraint(equalTo: red.view.trailingAnchor, constant: -16).isActive = true
+        redLabel.bottomAnchor.constraint(equalTo: red.view.bottomAnchor, constant: -304).isActive = true
+        
+        let blue = UIViewController()
+        let blueImage = UIImage(named: "blue")
+        let blueImageView = UIImageView(image: blueImage)
+        blue.view.addSubview(blueImageView)
+        blueImageView.translatesAutoresizingMaskIntoConstraints = false
+        blueImageView.leadingAnchor.constraint(equalTo: blue.view.leadingAnchor).isActive = true
```
Для каждой страницы "листалки" онбординга лучше создать отдельный подкласс `UIViewController`. Не нужно будет каждый раз создавать констрэйнты, вьюхи. Достаточно будет передать данные об отображении в пару строчек кода.

Lagowazzz replied:
Готово

Helvetii951 commented on Trackers/NewTrackers/Category/CategoryViewController.swift:
```diff
@@ -6,6 +6,8 @@ protocol CategoryViewControllerDelegate: AnyObject {
 
 final class CategoryViewController: UIViewController {
     
+    private var viewModel: CategoryViewModel?
+    
```
Для гибкости и тестируемости кода желательно объявлять вью-модель протоколом и как-нибудь инжектить.

[Неплохая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/).

Я бы еще сделал VM не опцианольной. Если она `nil`, в экране не будет смысла, значит что-то пошло не так.

Lagowazzz replied:
Почитаю обязательно

Helvetii951 commented on Trackers/NewTrackers/Category/CategoryViewController.swift:
```diff
@@ -61,6 +63,11 @@ final class CategoryViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
+        viewModel = CategoryViewModel(categoryStore: TrackerCategoryStore())
+        viewModel?.updateHandler = { [weak self] in
+            self?.tableView.reloadData()
+        }
```
Настройку бындингов стоит вынести в отдельную функцию. Так будет легче ориентироваться в коде.

Lagowazzz replied:
Готово

Helvetii951 commented on Trackers/SceneDelegate.swift:
```diff
@@ -1,14 +1,36 @@
+            let tabBarController = TabBarController()
+            window?.rootViewController = tabBarController
+        } else {
+            let onboardingViewController = OnboardingViewController()
+            onboardingViewController.onboardingCompletionHandler = {
+                appSettingsStore.setHasSeenOnboarding(true)
+                let tabBarController = TabBarController()
+                self.window?.rootViewController = tabBarController
+            }
+            window?.rootViewController = onboardingViewController
```
Клево, что перенес ответственность по навигации из контроллера онбординга👍

Lagowazzz replied:
😊

Helvetii951 commented on Trackers/NewTrackers/Category/CategoryViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+import Foundation
+
+final class CategoryViewModel {
+    
+    typealias Binding<T> = (T) -> Void
+    
```
Объявление биндинга можно вынести в отдельный файл. В теории он может пригодиться в других VM. Навигация по проекту будет понятнее.

Lagowazzz replied:
Done

Helvetii951 commented on Trackers/NewTrackers/Category/CategoryViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+    
+    private var categories: [TrackerCategory] = []
+    private let categoryStore: TrackerCategoryStoreProtocol
+    
+    var updateHandler: (() -> Void)?
+    
```
Тип биндинга объявлен, но не используется для реализации. Стоит применить объявление к `updateHandler`.

Lagowazzz replied:
Done

Lagowazzz committed:
Work on mistakes
Binding.swift for typelias binding
OnboardingSetupPageVC was created for OnboardingVC
